[C10D] Reimplement TCPStore wait timeout logic.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100594

Current TCPStore wait logic leaves the client socket in a bad state if waiting timesout.

This happens because all recv functions raise an exception on timeout and that's it.
The problem is that on timeout we need to unregister the wait.

We implement this with client side cancelation by adding a new CANCEL_WAIT instruction.

So, if no data arrives before the deadline, the client sends a CANCEL_WAIT command.
The server sends a WAIT_CANCELED response to that command, always.

This gets us down to the last issue, which is that there's a race between timeout'ing,
canceling the wait and the wait completing. The client needs to handle the server sending
a STOP_WAITING followed by a WAIT_CANCELED answer.

This ensures client and server state are synchronized regardless of whether the wait
timeouts or not.

kumpera committed:
[C10D] Reimplement TCPStore wait timeout logic.

Current TCPStore wait logic leaves the client socket in a bad state if waiting timesout.

This happens because all recv functions raise an exception on timeout and that's it.
The problem is that on timeout we need to unregister the wait.

We implement this with client side cancelation by adding a new CANCEL_WAIT instruction.

So, if no data arrives before the deadline, the client sends a CANCEL_WAIT command.
The server sends a WAIT_CANCELED response to that command, always.

This gets us down to the last issue, which is that there's a race between timeout'ing,
canceling the wait and the wait completing. The client needs to handle the server sending
a STOP_WAITING followed by a WAIT_CANCELED answer.

This ensures client and server state are synchronized regardless of whether the wait
timeouts or not.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Reimplement TCPStore wait timeout logic."


Current TCPStore wait logic leaves the client socket in a bad state if waiting timesout.

This happens because all recv functions raise an exception on timeout and that's it.
The problem is that on timeout we need to unregister the wait.

We implement this with client side cancelation by adding a new CANCEL_WAIT instruction.

So, if no data arrives before the deadline, the client sends a CANCEL_WAIT command.
The server sends a WAIT_CANCELED response to that command, always.

This gets us down to the last issue, which is that there's a race between timeout'ing,
canceling the wait and the wait completing. The client needs to handle the server sending
a STOP_WAITING followed by a WAIT_CANCELED answer.

This ensures client and server state are synchronized regardless of whether the wait
timeouts or not.

[ghstack-poisoned]

xw285cornell commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -1148,16 +1170,34 @@ void TCPStore::wait(
+  if (client_->receiveValueWithTimeout<detail::WaitResponseType>(
+          response, timeout)) {
+    if (response != detail::WaitResponseType::STOP_WAITING) {
+      TORCH_CHECK(false, "Stop_waiting response is expected");
```
will we hard stop here (if we receive STOP_WAIT we'll return, if we receive WAIT_CANCELED we'll fail the torch check, so whatever below this is not run?

kumpera replied:
Yes, exactly.

It would be a bug to receive WAIT_CANCELED since at this point the client has not sent a CANCEL_WAIT command yet.

H-Huang commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -1148,16 +1170,34 @@ void TCPStore::wait(
+    if (response != detail::WaitResponseType::STOP_WAITING) {
+      TORCH_CHECK(false, "Stop_waiting response is expected");
+    }
+    return;
```
i dont follow why this `return` is here, doesn't that mean CANCEL_WAIT doesn't get sent?

kumpera replied:
Yes, CANCEL_WAIT doesn't get sent. 
But it's not necessary to send it if the server already replied to stop waiting.
We only send it in the timeout path.

kumpera committed:
Update on "[C10D] Reimplement TCPStore wait timeout logic."


Current TCPStore wait logic leaves the client socket in a bad state if waiting timesout.

This happens because all recv functions raise an exception on timeout and that's it.
The problem is that on timeout we need to unregister the wait.

We implement this with client side cancelation by adding a new CANCEL_WAIT instruction.

So, if no data arrives before the deadline, the client sends a CANCEL_WAIT command.
The server sends a WAIT_CANCELED response to that command, always.

This gets us down to the last issue, which is that there's a race between timeout'ing,
canceling the wait and the wait completing. The client needs to handle the server sending
a STOP_WAITING followed by a WAIT_CANCELED answer.

This ensures client and server state are synchronized regardless of whether the wait
timeouts or not.

[ghstack-poisoned]

kumpera commented:
> Also can we add some tests here to just to verify the socket doesn't go bad when timeout happens?

We have some timeout tests, but I added another one covering the exact scenario this PR fixes.

H-Huang commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -1236,20 +1259,43 @@ void TCPStore::wait(
-  auto response = client_->receiveValue<detail::WaitResponseType>();
-  if (response != detail::WaitResponseType::STOP_WAITING) {
-    TORCH_CHECK(false, "Stop_waiting response is expected");
+    response = client_->receiveValue<detail::WaitResponseType>(); // ignore
```
what is the purpose for this second check? In what case are there two STOP_WAITING messages sent?

kumpera replied:
This covers the case where there's a race between client and server.

The server sends a STOP_WAITING response to the client at the same time the client sends a CANCEL_WAIT request.

In this case, we expect to get to receive two replies from the server .

H-Huang reviewed (approved):
overall lgtm, just had one question

kumpera committed:
Update on "[C10D] Reimplement TCPStore wait timeout logic."


Current TCPStore wait logic leaves the client socket in a bad state if waiting timesout.

This happens because all recv functions raise an exception on timeout and that's it.
The problem is that on timeout we need to unregister the wait.

We implement this with client side cancelation by adding a new CANCEL_WAIT instruction.

So, if no data arrives before the deadline, the client sends a CANCEL_WAIT command.
The server sends a WAIT_CANCELED response to that command, always.

This gets us down to the last issue, which is that there's a race between timeout'ing,
canceling the wait and the wait completing. The client needs to handle the server sending
a STOP_WAITING followed by a WAIT_CANCELED answer.

This ensures client and server state are synchronized regardless of whether the wait
timeouts or not.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Reimplement TCPStore wait timeout logic."


Current TCPStore wait logic leaves the client socket in a bad state if waiting timesout.

This happens because all recv functions raise an exception on timeout and that's it.
The problem is that on timeout we need to unregister the wait.

We implement this with client side cancelation by adding a new CANCEL_WAIT instruction.

So, if no data arrives before the deadline, the client sends a CANCEL_WAIT command.
The server sends a WAIT_CANCELED response to that command, always.

This gets us down to the last issue, which is that there's a race between timeout'ing,
canceling the wait and the wait completing. The client needs to handle the server sending
a STOP_WAITING followed by a WAIT_CANCELED answer.

This ensures client and server state are synchronized regardless of whether the wait
timeouts or not.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Reimplement TCPStore wait timeout logic."


Current TCPStore wait logic leaves the client socket in a bad state if waiting timesout.

This happens because all recv functions raise an exception on timeout and that's it.
The problem is that on timeout we need to unregister the wait.

We implement this with client side cancelation by adding a new CANCEL_WAIT instruction.

So, if no data arrives before the deadline, the client sends a CANCEL_WAIT command.
The server sends a WAIT_CANCELED response to that command, always.

This gets us down to the last issue, which is that there's a race between timeout'ing,
canceling the wait and the wait completing. The client needs to handle the server sending
a STOP_WAITING followed by a WAIT_CANCELED answer.

This ensures client and server state are synchronized regardless of whether the wait
timeouts or not.

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge
